3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/PiiPseudonymizerEnrichment.scala

```diff
@@ -0,0 +1,237 @@
+  def transformer(transformMap: TransformMap): TransformMap = transformMap ++ fieldList.map(_.transformer(transformMap)).reduce(_ ++ _)
+}
+
+/**
+ * Specifies a field in POJO and the strategy that should be applied to it.
+ * @param strategy the strategy that should be applied
+ * @param fieldName the field where the strategy will be applied
+ */
+final case class PiiPojo(strategy: PiiStrategy, fieldName: String) extends PiiField {
+  override def applyStrategy(fieldValue: String): String = strategy.scramble(fieldValue)
+}
+
+/**
+ * Specifies a strategy to use, a field (should be "contexts") where the JSON can be found, a schema criterion to
+ * discriminate which contexts to apply this strategy to, and a json path within the contexts where this strategy will
+ * be apllied (the path may correspond to multiple fields).
+ *
+ * @param strategy the strategy that should be applied
+ * @param fieldName the field in POJO where the json is to be found
+ * @param schemaCriterion the schema for which the strategy will be applied
+ * @param jsonPath the path where the strategy will be applied
+ */
+final case class PiiJson(strategy: PiiStrategy, fieldName: String, schemaCriterion: SchemaCriterion, jsonPath: String) extends PiiField {
+  implicit val json4sFormats = DefaultFormats
+
+  override def applyStrategy(fieldValue: String): String =
+    compact(render(parse(fieldValue).transformField {
+      case JField("data", contents) =>
+        ("data", contents.transform {
+          case ja: JArray =>
+            ja.transform {
+              case JObject(context) =>
+                val fields: List[JValue] = List("schema", "data").flatMap(context.toMap.get)
+                if (fields.size == 2 && SchemaKey.parse(fields(0).extract[String]).map(schemaCriterion.matches(_)).getOrElse(false))
+                  JObject(List(("schema", fields(0)), ("data", jsonPathReplace(fields(1)))))
+                else JObject(context)
+            }
+        })
+    }))
+
+  //Configuration for JsonPath
+  private val jsonPathConf =
+    Configuration.builder().options(JOption.SUPPRESS_EXCEPTIONS).jsonProvider(new JacksonJsonNodeJsonProvider()).build()
+
+  /**
+   * Replaces a value in the given context data with the result of applying the strategy that value.
+   *
+   */
+  private def jsonPathReplace(jValue: JValue): JValue = {
+    val objectNode      = JsonMethods.mapper.valueToTree[ObjectNode](jValue)
+    val documentContext = JJsonPath.using(jsonPathConf).parse(objectNode)
+    documentContext.map(
+      jsonPath,
+      new MapFunction {
+        override def map(currentValue: scala.Any, configuration: Configuration): AnyRef = currentValue match {
+          case s: String => strategy.scramble(s)
+          case a: ArrayNode =>
+            a.elements.asScala.map {
+              case t: TextNode     => strategy.scramble(t.asText())
+              case default: AnyRef => default
+            }
+          case default: AnyRef => default
```

Comment:
`currentValue` is `Any` and the fallback is on `AnyRef`, what happens on `AnyVal`?

Reply:
You cannot get AnyVal, given the interface of MapFunction.

Reply:
why isn't currentValue AnyRef then?